src/pkg/services/m365/custom/drive_item.go

```diff
@@ -0,0 +1,339 @@
+	return c.deleted
+}
+
+func (c *driveItem) GetRoot() interface{} {
+	return c.root
+}
+
+func (c *driveItem) GetFile() fileItemable {
+	return c.file
+}
+
+func (c *driveItem) GetParentReference() itemReferenceable {
+	return c.parentRef
+}
+
+// TODO(pandeyabs): Should we only support GETs?
+func (c *driveItem) SetParentReference(parent itemReferenceable) {
+	c.parentRef = parent
+}
+
+func (c *driveItem) GetCreatedBy() identitySetable {
+	return c.createdBy
+}
+
+func (c *driveItem) GetCreatedByUser() userable {
+	return c.createdByUser
+}
+
+func (c *driveItem) GetLastModifiedByUser() userable {
+	return c.lastModifiedByUser
+}
+
+func (c *driveItem) GetCreatedDateTime() *time.Time {
+	return &c.createdDateTime
+}
+
+func (c *driveItem) GetLastModifiedDateTime() *time.Time {
+	return &c.lastModifiedDateTime
+}
+
+func (c *driveItem) GetAdditionalData() map[string]interface{} {
+	return c.additionalData
+}
+
+type (
+	malwareable interface {
+		GetDescription() *string
+	}
+	fileItemable interface {
+		GetMimeType() *string
+	}
+	itemReferenceable interface {
+		GetPath() *string
+		GetId() *string
+		GetName() *string
+		GetDriveId() *string
+	}
+	identitySetable interface {
+		GetUser() identityable
+	}
+	identityable interface {
+		GetAdditionalData() map[string]interface{}
+	}
+	userable interface {
+		GetId() *string
+	}
+)
+
+// Concrete implementations
+
+var _ malwareable = &malware{}
+
+type malware struct {
+	description string
+}
+
+func (m *malware) GetDescription() *string {
+	return &m.description
+}
+
+var _ fileItemable = &fileItem{}
+
+type fileItem struct {
+	mimeType string
+}
+
+func (f *fileItem) GetMimeType() *string {
+	return &f.mimeType
+}
+
+var _ itemReferenceable = &parentRef{}
+
+type parentRef struct {
+	path    string
+	id      string
+	name    string
+	driveID string
+}
+
+func (pr *parentRef) GetPath() *string {
+	return &pr.path
+}
+
+func (pr *parentRef) GetId() *string {
+	return &pr.id
+}
+
+func (pr *parentRef) GetName() *string {
+	return &pr.name
+}
+
+func (pr *parentRef) GetDriveId() *string {
+	return &pr.driveID
+}
+
+var _ identitySetable = &identitySet{}
+
+type identitySet struct {
+	identity identityable
+}
+
+func (iis *identitySet) GetUser() identityable {
+	return iis.identity
+}
+
+var _ identityable = &identity{}
+
+type identity struct {
+	additionalData map[string]interface{}
+}
+
+func (iu *identity) GetAdditionalData() map[string]interface{} {
+	return iu.additionalData
+}
+
+var _ userable = &user{}
+
+type user struct {
+	id string
+}
+
+func (u *user) GetId() *string {
+	return &u.id
+}
+
+// TODO(pandeyabs): This is duplicated from collection/drive package.
+// Move to common pkg
+var downloadURLKeys = []string{
+	"@microsoft.graph.downloadUrl",
+	"@content.downloadUrl",
+}
+
+func ToLiteDriveItemable(item models.DriveItemable) LiteDriveItemable {
+	if item == nil {
+		return nil
+	}
+
+	di := &driveItem{
+		id:                   strings.Clone(ptr.Val(item.GetId())),
```

Comment:
Is it necessary to clone the string?  ptr.Val should already produce a copy of the value.

Reply:
True, but it'll add unnecessary mem copies in most places. Since memcpy has a cpu cost, i think we should avoid it unless really needed. We could however expose another function `ptr.ValCopy` or similar with `Clone()` in it.

Reply:
Okay, sounds good.  I don't think we need a `ValCopy`.  It's too generic a func, anyway, and would have lots of implications beyond just string value handling.  If we find ourselves going back to it often, `common/str` is probably the right owner.  `str.PtrVal(s)` or somesuch.